scripts/regression/utils.py

```diff
@@ -1,13 +1,42 @@
 import glob
 import json
+import tqdm
 import os
 from typing import Tuple
 
+from spacer import config
 from spacer.data_classes import ImageLabels
 from spacer.messages import DataLocation
 from spacer.train_classifier import trainer_factory
 
 
+def cache_local(source_root: str,
+                image_root: str,
+                export_name: str,
+                source_id: int) -> None:
```

Comment:
@qiminchen : do you want to add a boolean to say whether to cache the pre-extracted features? I don't think we need to download them for all regression tests (i.e. the new you you added)

Reply:
hmm what do you mean, i didnt quite get it..

Reply:
In the `train_classifier.py` regression script we need to download the features from S3 since we don't re-extract them. But we don't need the images. In `efficientnet_extractor.py` the opposite is true: we need the images but not the feature files. So I'm suggesting adding options to this caching script to just fetch what you actually need. (In all cases the metadata an labels can be fetched -- they are so small)

Reply:
Actually, looking closer at the code now I see what you did. In the `ExtractFeatures` you fetch the images directly from S3 and write to local, so you both down-load and extract at the same time. I think that is fine. But wouldn't then `utils.cache_local` over-write the extracted features?

Reply:
I think the code will be cleaner and run faster if you cache first and the do extraction and training locally.

Reply:
> Actually, looking closer at the code now I see what you did. In the `ExtractFeatures` you fetch the images directly from S3 and write to local, so you both down-load and extract at the same time. I think that is fine.

Yes, In the `ExtractFeatures` I read the images directly from S3 and only write extracted features to local, in other words, only the extracted features will be cached to local.

> But wouldn't then `utils.cache_local` over-write the extracted features?

And no, after feature extraction, we cache the data to local, since we haven't add the boolean to specify whether to fetch feature or not, it should fetch the features from s3, but since we just extract the features (have the exact same name as what they are named in S3), features from s3 will be skipped and only metadata and labels will be fetched. So in `efficientnet_extractor.py`, `utils.cache_local` will not over-write the extracted features.
```
    for key in tqdm.tqdm(img_keys):
        _, filename = key.name.split('images')
        local_path = os.path.join(image_root, filename.lstrip('/'))
        if not os.path.exists(local_path):
            key.get_contents_to_filename(local_path)
```
In both cases, images will not be cached to local but i think its good to add a boolean to `utils.cache_local` since in `efficientnet_extractor.py`, features will not be fetched.

Reply:
> I think the code will be cleaner and run faster if you cache first and the do extraction and training locally. This allows you to remove all s3 stuff in `ExtractFeatures` also.

so you think i should cache the images to local? because right now i didn't cache the images to local but only extracted features.

Reply:
Thanks for the explanation. I forgot there was the check to download files that already existed. Ok, so the current code should work fine then. 

You can choose - I think the code will be cleaner if you manage all AWS fetching in the `cache_local` and then operate locally from then on. And I also suspect it will run a bit faster.

Reply:
yes, the code will be cleaner for `efficientnet_extractor.py` if we cache images, metadata and labels to local first and then operate locally from them on (extracting features and training). The only tradeoff is that the image will be downloaded and it takes up some space. I'm okay with both,  which one do you prefer? It's easy to modify the code.

Reply:
Let's do the change then. I think it will be easier to follow what is going on.

Reply:
okay, working on it

Reply:
okay, all AWS fetching are moved to the `cache_local`, but I didn't feel the whole procedure runs faster.
Current: download images to local + extract features
Previous: read images from S3 + extract features
I didn't test the time spend on downloading the images vs. reading images from S3, which seems to be pretty similar. Besides, downloading the images to the local takes up much space especially for large sources, each original image is in mb